source/AI.cpp

```diff
@@ -1194,40 +1194,65 @@ void AI::MoveEscort(Ship &ship, Command &command) const
 		Refuel(ship, command);
 	else if(!parentIsHere && !isStaying)
 	{
-		// If this ship has already refuelled, and its parent has left the
-		// system, no need to land on a planet again.
-		if(ship.GetTargetStellar() && ship.Fuel() == 1. && ship.GetTargetStellar()->GetPlanet()
-				&& !ship.GetTargetStellar()->GetPlanet()->IsWormhole())
-			ship.SetTargetStellar(nullptr);
+		if(ship.GetTargetStellar())
+		{
+			const Planet *targetPlanet = ship.GetTargetStellar()->GetPlanet();
+			if(!targetPlanet || !targetPlanet->CanLand(ship))
+				ship.SetTargetStellar(nullptr);
+			// If this ship has already refuelled, and its parent has left the
+			// system, no need to land on a planet again.
+			else if(!targetPlanet->IsWormhole() && ship.Fuel() == 1.)
+				ship.SetTargetStellar(nullptr);
+		}
 		
-		// Check whether the ship has a target system and is able to jump to it,
-		// and that the targeted stellar object can be landed on.
-		bool hasJump = (ship.GetTargetSystem() && ship.JumpFuel(ship.GetTargetSystem()));
-		if(ship.GetTargetStellar()
-				&& !(ship.GetTargetStellar()->GetPlanet() && ship.GetTargetStellar()->GetPlanet()->CanLand(ship)))
-			ship.SetTargetStellar(nullptr);
-		if(!hasJump && !ship.GetTargetStellar())
+		if(!ship.GetTargetStellar() && !ship.GetTargetSystem())
 		{
-			// If we're stranded and haven't decided where to go, figure out a
-			// path to the parent ship's system.
+			// Figure out a path to the parent ship's system and check whether the
+			// ship should refuel, land on a wormhole or jump to the next system.
 			DistanceMap distance(ship, parent.GetSystem());
 			const System *from = ship.GetSystem();
-			const System *to = distance.Route(from);
-			for(const StellarObject &object : from->Objects())
-				if(object.GetPlanet() && object.GetPlanet()->WormholeDestination(from) == to)
-				{
-					ship.SetTargetStellar(&object);
+			double requiredFuel = 0.;
+			while(from != parent.GetSystem())
+			{
+				const System *to = distance.Route(from);
+				bool useWormhole = false;
+				for(const StellarObject &object : from->Objects())
+					if(object.GetPlanet() && object.GetPlanet()->WormholeDestination(from) == to)
+					{
+						useWormhole = true;
+						if(from == ship.GetSystem())
+							ship.SetTargetStellar(&object);
+						break;
+					}
+				
+				if(!systemHasFuel || !hasFuelCapacity)
 					break;
+				
+				// Check how much fuel is required to reach the next refuel system.
+				if(!useWormhole)
+				{
+					if(from->Links().count(to))
+						requiredFuel += ship.HyperdriveFuel();
+					else if(from->Neighbors().count(to))
+						requiredFuel += ship.JumpDriveFuel();
+					else
+					{
+						// How do I get to the next system? Refuel just in case...
+						requiredFuel = ship.Attributes().Get("fuel capacity");
```

Comment:
Is a ship that can't reach the parent's system going to refuel after every jump it makes?

Reply:
This is a failsafe for when this code doesn't know how DistanceMap decided you can reach the next system in the route. It should never happen for now, but might happen when DistanceMap is modified.

Reply:
If the route is empty it throws nullptr back, so you will fall down to this, e.g. non jump escorts capped in FW space and you jump to Coalition space.